### DISCUSSION TITLE
Is it possible to continue processing after sending a respone?

### DISCUSSION THREAD

**Discussion by sinn1** [vercel/next.js#12573] [Help]
I'm working on a slack command and they recommend to send an Acknowledgement response immediately: https://api.slack.com/interactivity/handling#responses#acknowledgment_response

I was expecting something along the lines of this to work:

```
(req, res) => {
  res.statusCode = 200;
  res.json({ text: 'Ok!' });

  doSomeWork();
}

```

This works when running locally, but when deployed to zeit now (vercel) it seems to exit after res.json() - I've added logging before and after to verify this. 

Is there a recommended pattern for achieving this behavior? Alternatively, is there a recommended way to chain serverless function calls? Is this recommended:

```
(req, res) => {
  res.statusCode = 200;
  res.json({ text: 'Ok!' });

 axios.post(`${process.env.VERCEL_URL}/api/doSomeWork, data)
}

```

Thanks!

**Comment by brianlovin** [comment]
My understanding is that `res.json` calls `res.send` under the hood, and `res.send` calls `res.end`. So yes, it does seem like you'd need to "pass through" the payload to some handler. I did a bit of googling around because I find this to be an interesting problem that I might encounter in the future with API routes; it seems like in the short term you could basically do what you described:

- have a message receiver endpoint
- the receiver dispatches an action to a separate event handler endpoint
- the receiver responds with acknowledgement

One thing to consider is you might want to trigger the `doSomeWork` event before calling `res.json`

**Comment by mohouyizme** [comment]
I believe as long as you're calling an asynchronous function it won't affect the response time

![carbon (7)](https://user-images.githubusercontent.com/23297439/81227537-1bc9ab80-8fdc-11ea-8cc3-95aed49de9e5.png)

**Reply by sinn1** [reply]
I've tested this and it kind of works, with a bit of a tweak. I've had to add a short timeout:

```
 async (req, res) => {

  // call async function without await
  doSomeWork();

  // await a short timeout 
  await waitInMilliseconds(100);

  // send response
  res.json({ text: 'Ok!' });
}
```

Something strange I've noticed, however, is that if you repeatedly call the endpoint you intermittently receive this response - it feels like every 1-3 requests returns this.
```
An error occurred with this application.

NO_STATUS_CODE_FROM_FUNCTION
```

**Reply by oliverox** [reply]
Are you sure it's not the 100ms delay that is allowing `doSomeWork()` to execute? Did you try making `doSomeWork()` run for say 5000ms with that scenario?

**Comment by brianlovin** [comment]
> I believe as long as you're calling an asynchronous function it won't affect the response time

Interesting - what if `doSomeWork` takes a long time - will the function keep running in the background?

**Reply by mohouyizme** [reply]
Exactly!

**Comment by enricoschaaf** [comment]
#11661 I had a similar case, as far as I understand it's not possible. Maybe you could call another serverless function without awaiting the response and let this function execute your longer task.

**Comment by lkbr** [comment]
If you're hosting with Vercel, this [PR](https://github.com/zeit/docs/pull/1821) might bring some clarity:

> This means that background tasks, also known as "fire-and-forget" is not supported. Once the function returns the response payload, it stops processing including any pending background tasks.

**Reply by sinn1** [reply]
This sounds like it should be the correct answer. From the official docs: https://docs-git-add-fire-and-forget.zeit.now.sh/docs/v2/platform/limits?query=fire-and-forget#streaming-responses

However, @mohouyizme's answer [above](https://github.com/zeit/next.js/discussions/12573#discussioncomment-10509) is a workaround. (See my reply to that answer for a caveat).

**Comment by doxylee** [comment]
EDIT: This doesn't seem to work.. It allows code to run after res.json, but it also keeps the response from being sent until the code is run, which makes this solution pointless. Can anyone else confirm this?

~~There's some code from @sentry/nextjs that keeps lambda function alive so that sentry can send error logs after the response.~~

~~https://github.com/getsentry/sentry-javascript/blob/ad37109568479bb8459b76f6bdfe0c0bd54e66dd/packages/nextjs/src/utils/withSentry.ts~~

~~Based off it, I created a wrapper function that keeps lambda process running until the api handler is resolved.
I'm using it with the original `withSentry`, so I left out protection against scope bleeding, nextjs complaining res.end() not being called, etc
I deployed it and it seems to work, but I didn't test it throughly. There might be some bugs :(~~

``` Typescript
export function withFireAndForget(
    handler: (req: NextApiRequest, res: NextApiResponse) => Promise<void>
) {
    return async (req: NextApiRequest, res: NextApiResponse) => {
        let promiseResolve: () => void = ()=>{throw Error("promiseResolved was not reassigned!")}
        const finishedPromise = new Promise<void>((resolve) => {
            promiseResolve = resolve;
        });
        const origEnd = res.end;
        res.end = async function newEnd(this: NextApiResponse, ...args: unknown[]) {
            await finishedPromise;
            //@ts-ignore
            return origEnd.call(this, ...args);
        };

        const result = await handler(req, res);
        promiseResolve();
        return result;
    };
}
```

~~Anyway, hope it helps.~~

**Reply by notturingtested** [reply]
I had the same thing happen. Did you come up with a solution to this? Stripe's payment succeeded webhook has a short timeout, and I need to write to the DB and send an email!

**Reply by doxylee** [reply]
@notturingtested  No, I think it's impossible due to infra level limitation. You should call another api without awaiting for a response, or use a queueing system like AWS SQS.

**Comment by jodik** [comment]
Calling an async function before calling res.json() works on Cloud Run with CPU Always On

**Reply by l4b4r4b4b4** [reply]
latest canary release makes it possible. Just found out after one week of looking for a solution!

**Reply by christopherliedtke** [reply]
do you have any documentation about this?

**Reply by l4b4r4b4b4** [reply]
you mean me? just do `pnpm add next@canary`, `yarn add next@canary` or `npm add next@canary` depending on the package manager you use in your respective project.
If you got further questions, come to Next.js discord server. I am online there ;)

> do you have any documentation about this?
